crates/esplora/src/blocking_ext.rs

```diff
@@ -68,250 +56,368 @@ pub trait EsploraExt {
+    anchors: &BTreeSet<(A, Txid)>,
+    mut update_blocks: BTreeMap<u32, BlockHash>,
+) -> Result<local_chain::Update, Error> {
+    let update_tip_height = update_blocks
+        .keys()
+        .last()
+        .copied()
+        .expect("must atleast have one block");
+
+    // We want to have a corresponding checkpoint per height. We iterate the heights of anchors
+    // backwards, comparing it against our `local_tip`'s chain and our current set of
+    // `update_blocks` to see if a corresponding checkpoint already exists.
+    let anchor_heights = anchors
+        .iter()
+        .map(|(a, _)| a.anchor_block().height)
+        // filter out duplicate heights
+        .filter({
+            let mut prev_height = Option::<u32>::None;
+            move |h| match prev_height.replace(*h) {
+                None => true,
+                Some(prev_h) => prev_h != *h,
             }
+        })
+        // filter out heights that surpass the update tip
+        .filter(|h| *h <= update_tip_height)
+        .rev();
+
+    // We keep track of a checkpoint node of `local_tip` to make traversing the linked-list of
+    // checkpoints more efficient.
+    let mut curr_cp = local_tip.clone();
 
-            if let Some(last_active_index) = last_active_index {
-                last_active_indexes.insert(keychain, last_active_index);
+    for h in anchor_heights {
+        if let Some(cp) = curr_cp.query_from(h) {
+            curr_cp = cp.clone();
+            if cp.height() == h {
+                // blocks that already exist in checkpoint linked-list is also stored in
+                // `update_blocks` because we want to keep higher checkpoints of `local_chain`
+                update_blocks.insert(h, cp.hash());
+                continue;
             }
         }
-
-        Ok((graph, last_active_indexes))
+        if let btree_map::Entry::Vacant(entry) = update_blocks.entry(h) {
+            entry.insert(client.get_block_hash(h)?);
+        }
     }
 
-    fn sync(
-        &self,
-        misc_spks: impl IntoIterator<Item = ScriptBuf>,
-        txids: impl IntoIterator<Item = Txid>,
-        outpoints: impl IntoIterator<Item = OutPoint>,
-        parallel_requests: usize,
-    ) -> Result<TxGraph<ConfirmationTimeHeightAnchor>, Error> {
-        let mut graph = self
-            .full_scan(
-                [(
-                    (),
-                    misc_spks
-                        .into_iter()
-                        .enumerate()
-                        .map(|(i, spk)| (i as u32, spk)),
-                )]
-                .into(),
-                usize::MAX,
-                parallel_requests,
-            )
-            .map(|(g, _)| g)?;
-
-        let mut txids = txids.into_iter();
+    Ok(local_chain::Update {
+        tip: CheckPoint::from_block_ids(
+            update_blocks
+                .into_iter()
+                .map(|(height, hash)| BlockId { height, hash }),
+        )
+        .expect("must be in order"),
```

Comment:
Isn't it much more straightforward to just extend the existing chain by pushing onto the `local_tip` one from the point of agreement. Why are we constructing an entirely new one.

Reply:
Because we mark the update to introduce older checkpoints, and in which case we will iterate the whole chain.

Reply:
Hmm don't we know if the thing actually does introduce older checkpoints? We don't have to fix `introduce_older_blocks` to `true`. Also building from the existing checkpoint will allow us to stop early if the `Arc` pointers match.

IIRC `introduce_older_blocks` was needed so we could set it to false for block-by-blocking syncing where if it were assumed to be true and some edge case happened then the `Arc` pointers wouldn't match and the update would be computationally intensive.

We could probably overcome this with a block `xor` state accumulator thingo that you had in the other PR. So I'd work on doing this in the most straightforward way and then probably remove the whole `introduce_older_blocks` thing in favor of block xor to detect if something has changed instead of relying on `Arc` pointers.

Reply:
I see what you are saying. I think I did try to construct the update from the old tip, but it turned out more complicated than I expected? We have to traverse the old chain to find a height just lower than our update's earliest height. I think I can try look at this problem again to see if I can simplify it.

Reply:
Perhaps it would feel less complicated if you made an `insert(height, hash)` method on `CheckPoint` which consumed `self` and returned a new chain with that height and hash inserted at the right point.

I think we should remove the whole `introduce_older_blocks` thing in this PR and just assume it's true and make it faster later. I think we know how to do that now.

Reply:
I think removing the `introduce_older_blocks` thing in this PR is a bad idea. It will make block-by-block chain sources substantially slower (and they are already being used).

I forgot that the `merge_chains` method already has an optimization to stop when we reach a checkpoint with a matching pointer. This should be good enough to break early for spk-based chain sources.

However, for block-by-block chain sources, we aren't given the opportunity to re-introduce local checkpoints to the `Emitter`.

I think we can work to remove `introduce_older_blocks` in a separate PR.

Reply:
Is it important to construct the update from the older checkpoint? Is the intention to simplify the logic? I don't think it will make things simpler.

Reply:
I made a PR to this branch to show how much simpler things could be: https://github.com/evanlinjin/bdk/pull/9

Reply:
On removing introduce_older_blocks I'm inclined to do it now. Make it right, then make it fast. I also think it will be fast anyway except in the unlikely case of a re-org (Arc's pointers will match so won't slow down in practice unless I'm forgetting how block by block syncing works). People using `alpha` things should be ready for unexpected slowdown. They don't have to update to the latest `alpha` until it's fixed.

Reply:
daf588f016ec3118c875db8ed6b55fa03683f0f6

Let me know what u think